### PR TITLE
Feature/issue 13

### PR DESCRIPTION
**Pull request by jlanka11566**
Allows use of wrapper command to download terraform binaries

### CODE REVIEW

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+    /*
+     * Here we check for command line arguments (if any) and update the properties file
+     */
+    final File propFile    = new File(System.getProperty("user.dir") + "\\.tf\\terraform-maven.properties");
+    final File newPropFile = new File(System.getProperty("user.dir") + "\\.tf\\terraform-maven.properties2");
+    final boolean propExists = propFile.exists();
+
+    if (propExists) {
+      try {
```
[the remark this answers is not in the document]
Comment (Clayton7510) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:118]: Here's an example of reading into a java Properties object from a properties file.
```java
Properties properties = new Properties();
properties.setProperty("releaseDir", releaseDir);
...
try (InputStream fis = new FileInputStream(myPropertiesFile) {
   properties.load(fis);
} catch (IOException ioe) {
  // handle the exception here
}
```

Here's an example of writing to a properties file from a Java Properties object.
```java
try (OutputStream fos = new FileOutputStream(myPropertiesFile) {
   properties.store(fos, "Terraform Wrapper Properties");
} catch (IOException ioe) {
  // handle the exception here
}
```

File: tf-build-tools/tf-maven-plugin/src/test/java/com/deliveredtechnologies/maven/terraform/mojo/TerraformWrapperTest.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package com.deliveredtechnologies.maven.terraform.mojo;
```
Comment (Clayton7510) [inline tf-build-tools/tf-maven-plugin/src/test/java/com/deliveredtechnologies/maven/terraform/mojo/TerraformWrapperTest.java:1]: Think about how you might test this.

File: examples/tf-s3/pom.xml [hunk 3]
```diff
@@ -156,6 +156,7 @@
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-compiler-plugin</artifactId>
       </plugin>
+
```
[the remark this answers is not in the document]
Comment (jlanka11566) [reply examples/tf-s3/pom.xml:159]: Removed space

File: tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/io/Executable.java [hunk 4]
```diff
@@ -14,3 +14,4 @@
 
   public void setLogger(Logger logger);
 }
+
```
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-cmd-api/src/main/java/com/deliveredtechnologies/io/Executable.java:17]: Removed 1 additional line from bottom of file

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Scanner;
+
+
+@Mojo(name = "wrapper")
```
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:19]: Moved @Mojo(name = "wrapper") to line 24 directly above the class declaration

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+/*
```
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:20]: need confirmation

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+ * These are all the possible input parameters the user can supply with
+ * the -D switch
+ *
+ */
+public class Wrapper extends TerraformMojo<String> {
+  /*
+   * These are all the possible input parameters the user can supply with
+   * the -D switch
+   */
+  @Parameter( property = "distributionSite")
```
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:30]: Need confirmation

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  protected MojoExecution mojoExecution;
+
+
+  @Override
+
+  public void execute() throws MojoExecutionException, MojoFailureException {
+    /*
+     * Here we determine the OS in order to add that to the prop file
+     */
+    final String os_name = System.getProperty("os.name");
```
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:55]: Need confirmation

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+      } else {
+        inreleaseOS = "linux";
+      }
+    }
+
+    /*
+     * Here we declare all the final constants which represent the files to be
+     * copied from the jar file to the users project
+     */
+    final String fileName1 = "tf/tfw";
```
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:69]: 1) So how do I determine which to make final?
2) Great idea I will implement that!

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    final String fileName2 = "tf/tfw.cmd";
+    final String fileName3 = "tf/tfw.ps1";
+    final String fileName4 = "tf/terraform-maven.properties";
+    final URL f1 = (this.getClass().getClassLoader().getResource(fileName1));
+    final URL f2 = (this.getClass().getClassLoader().getResource(fileName2));
+    final URL f3 = (this.getClass().getClassLoader().getResource(fileName3));
+    final URL f4 = (this.getClass().getClassLoader().getResource(fileName4));
+    final File dest1 = new File(System.getProperty("user.dir") + "\\.tf\\tfw");
```
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:77]: Thanks I will implement this

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:118]: Thanks I will implement this

File: tf-build-tools/tf-maven-plugin/src/test/java/com/deliveredtechnologies/maven/terraform/mojo/TerraformWrapperTest.java (hunk 2 above)
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/test/java/com/deliveredtechnologies/maven/terraform/mojo/TerraformWrapperTest.java:1]: I will speak to you at 2 PM for some guideance

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:30]: Added lines between each @Parameter directive

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:55]: Added final for String os_name

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:69]: Removed final for all but os_name
Implemented array for 4 file names

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java:77]: Added this logic and removed old logic/variables as suggested

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java]: I implemented the Properties class but ran into 2 issues:
1) The properties file is written in DOS format which breaks the scripts
2) The line with the URL has the : escaped which also breaks the scripts

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java]: Modified comments as requested

File: tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/main/java/com/deliveredtechnologies/maven/terraform/mojo/Wrapper.java]: Implemented Properties class to manipulate properties file as requested

File: tf-build-tools/tf-maven-plugin/src/test/java/com/deliveredtechnologies/maven/terraform/mojo/TerraformWrapperTest.java (hunk 2 above)
Comment (jlanka11566) [reply tf-build-tools/tf-maven-plugin/src/test/java/com/deliveredtechnologies/maven/terraform/mojo/TerraformWrapperTest.java:1]: @Clayton7510  I pushed up the WrapperTest.java plus the changes you showed me to Wrapper.java.  Please let me know if any more changes need to be made. Thanks.

**Pull request merged**